modules/02_Theory.md

```diff
@@ -0,0 +1,371 @@
+
+### Module resolution is host-defined
+
+While the ECMAScript specification defines how to parse and interpret `import` and `export` statements, it leaves module resolution up to the host. If you’re creating a hot new JavaScript runtime, you’re free to create a module resolution scheme like:
+
+```ts
+import monkey from "🐒"; // Looks for './eats/bananas.js'
+import cow from "🐄";    // Looks for './eats/grass.js'
+import lion from "🦁";   // Looks for './eats/you.js'
+```
+
+and still claim to implement “standards-compliant ESM.” Needless to say, TypeScript would have no idea what types to assign to `monkey`, `cow`, and `lion` without built-in knowledge of this runtime’s module resolution algorithm. Just as `module` informs the compiler about the host’s expected module format, `moduleResolution`, along with a few customization options, specify the algorithm the host uses to resolve module specifiers to files.
+
+> #### Why doesn’t TypeScript delegate module resolution to third-party plugins?
+>
+> [TODO]
+>
+> * security boundary
+> * performance
+> * auto-imports and path completions
+> * chaos and conflicts
+
+The available `moduleResolution` options are:
+
+- **`classic`**: TypeScript’s oldest and most inscrutable module resolution mode, this is unfortunately the default when `module` is set to anything other than `commonjs`, `node16`, or `nodenext`. It does not represent any real-world module resolver and should never be used. It was kept around only for backwards compatibility.
+- **`node10`**: Formerly known as `node`, this is the unfortunate default when `module` is set to `commonjs`. It’s a pretty good model of Node versions older than v12, and sometimes it’s a passable approximation of how most bundlers do module resolution. It supports looking up packages from `node_modules`, loading directory `index.js` files, and omitting `.js` extensions in relative module specifiers. Because Node v12 introduced different module resolution rules for ES modules, though, it’s a very bad model of modern versions of Node. It should not be used for new projects.
+- **`node16`**: This is the counterpart of `--module node16` and is set by default with that `module` setting. Node v12 and later support both ESM and CJS, each of which uses its own module resolution algorithm. In Node, module specifiers in import statements and dynamic `import()` calls are not allowed to omit file extensions or `/index.js` suffixes, while module specifiers in `require` calls are. This module resolution mode understands and enforces this restriction where necessary, as determined by the [module detection rules](#module-kind-detection) instated by `--module node16`. (For `node16` and `nodenext`, `module` and `moduleResolution` go hand-in-hand: setting one to `node16` or `nodenext` while setting the other to something else has unsupported behavior and may be an error in the future.)
```

Comment:
`instated` is either a typo or a word I don't know

Reply:
Hmm, I guess I know reinstate. I thought it would be "enstate" or "instantiate".